src/electos/ballotmaker/data/models.py

```diff
@@ -0,0 +1,220 @@
+"""Ballot data models."""
+
+from dataclasses import dataclass
+from enum import Enum
+from functools import cached_property
+from typing import List, Union
+
+
+# --- Type assertions
+#
+# Python dataclasses don't check field types at runtime, but we want to avoid
+# errors.
+
+
+def _check_type(instance, field, type_):
+    """Raise 'TypeError' if 'instance.field' isn't of class 'type'."""
+    value = getattr(instance, field)
+    if not isinstance(value, type_):
+        raise TypeError(
+            f"Field '{field}' is not of type '{type_.__name__}': {value}"
+        )
+
+
+def _check_type_hint(instance, field, type_):
+    """Raise 'TypeError' if 'instance.field' isn't of type model 'type'."""
+    value = getattr(instance, field)
+    if not isinstance(value, type_):
+        raise TypeError(
+            f"Field '{field}' is not of type '{type_._name}': {value}"
+        )
+
+
+def _check_type_list(instance, field, type_):
+    """Raise 'TypeError' if 'instance.field' contents aren't of type 'type'."""
+    values = getattr(instance, field)
+    if not all(isinstance(value, type_) for value in values):
+        raise TypeError(
+            f"Values in field '{field}' are not all of type '{type_.__name__}': {values}"
+        )
+
+
+# --- Contest model base type
+
+@dataclass
+class ContestData:
+
+    """Shared data for contests."""
+
+    id: str
+    type: str
+    title: str
+    district: str
+
+    def __post_init__(self):
+        _check_type(self, "id", str)
+        _check_type(self, "type", str)
+        _check_type(self, "title", str)
+        _check_type(self, "district", str)
+
+
+class ContestType(Enum):
+
+    BALLOT_MEASURE = "ballot measure"
+    CANDIDATE = "candidate"
+
+
+# --- Ballot contest data models
+
+
+@dataclass
+class BallotChoiceData:
+
+    """Data for ballot measure selections."""
+
+    id: str
+    choice: str
+
+    def __post_init__(self):
+        _check_type(self, "id", str)
+        _check_type(self, "choice", str)
+
+
+@dataclass
+class BallotMeasureContestData(ContestData):
+
+    """Data for ballot measure contests."""
+
+    text: str
+    choices: List[BallotChoiceData]
+
+    def __post_init__(self):
+        super().__post_init__()
+        _check_type(self, "text", str)
+        _check_type_hint(self, "choices", List)
+        self.choices = [BallotChoiceData(**_) for _ in self.choices]
```

Comment:
Maybe if I just see this construct enough times it will make more sense. It seems to be clearer as a one-liner. I understand that you're unpacking a dict to a listf, but it's not clear what you get as a result. A list of strings? What happens to the readability of the line if you use a word instead of "_"?

By the way, this seems like a good use of `__post_init__()`